src/utils/__tests__/safeRename.spec.js

```diff
@@ -0,0 +1,79 @@
+import childProcess from 'child_process';
+import fsPromises from 'fs/promises';
+import fs from 'fs';
+import os from 'os';
+import { join, resolve } from 'path';
+import { remove as extraRemove } from 'fs-extra';
+
+import safeRename from '../safeRename.js';
+
+const assetsDir = resolve('./src/utils/__tests__/assets/safeRename');
+
+/* input tar file contents:
+ * rename1.txt
+ * a/
+ *   a1.txt
+ *   a2.txt
+ *   b/
+ *     b1.txt
+ */
+
+const mockFunc = jest.fn().mockImplementation(() => {
+  throw new Error('EXDEV: cross-device link not permittted');
+});
+
+fsPromises.rename = mockFunc;
+
+function fileExists(path) {
+  try {
+    fs.accessSync(path, fs.constants.F_OK);
+
+    return true;
+  } catch (_) {
+    return false;
+  }
+}
+
+describe('safeRename', () => {
+  let targetDir;
+
+  beforeEach(() => {
+    childProcess.execFileSync('tar', ['xf', join(assetsDir, 'safeRename.tar')], { cwd: assetsDir });
```

Comment:
I used an archive so I could easily create subtrees with enough complexity to test files got moved completely and correctly (e.g. moving `SRC/a` to `DEST/new-a` creates `DEST/new-a` and moves the contents of `SRC/a` to `DEST/new-a` but there is no `DST/new-a/a`. And the advantage of a tar file is that you don't have to worry about restoring files in the archive.

When I added a second test I noticed failures from the command-line but not the debugger. That's because when debugging I'm running only one test at a time. On the command-line you could have two separate tests that both try to move `SRC/a` at the same time. By `tar x`ing into a unique temp dir, that avoids that problem.